### DISCUSSION TITLE
GPU RAM requirements for training

### DISCUSSION THREAD

**Discussion by DSA101** [ultralytics/yolov5#5777] [Q&A] [answered]
How much GPU RAM is needed to train yolov5 s, m, l, x models with default parameters?

I want to understand whether I'll be able to train a high quality model on my laptop (RTX 3050, 4GB) or desktop (RTX 3060, 12GB) or it's way too little and I need to look for something more exotic. 

Similar RAM question for inference. 

Thanks!

**Comment by glenn-jocher** [comment, marked as answer]
@DSA101 👋 Hello! Thanks for asking about **CUDA memory issues**. YOLOv5 🚀 can be trained on CPU, [single-GPU](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data), or [multi-GPU](https://github.com/ultralytics/yolov5/issues/475). When training on GPU it is important to keep your batch-size small enough that you do not use all of your GPU memory, otherwise you will see a CUDA Out Of Memory (OOM) Error and your training will crash. You can observe your CUDA memory utilization using either the `nvidia-smi` command or by viewing your console output:

<img width="800" alt="Screenshot 2021-05-28 at 12 19 51" src="https://user-images.githubusercontent.com/26833433/131122318-8141d1ce-13de-44bf-bbdc-487bad442ff0.png">

### CUDA Out of Memory Solutions

If you encounter a CUDA OOM error, the steps you can take to reduce your memory usage are:

- Reduce `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s
- Train with [multi-GPU](https://github.com/ultralytics/yolov5/issues/475) at the same `--batch-size`
- Upgrade your hardware to a larger GPU
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### AutoBatch

You can use [YOLOv5 AutoBatch](https://github.com/ultralytics/yolov5/pull/5092) (NEW) to find the best batch size for your training by passing `--batch-size -1`. AutoBatch will solve for a 90% CUDA memory-utilization batch-size given your training settings. AutoBatch is experimental, and only works for Single-GPU training. It may not work on all systems, and is not recommended for production use.

<img width="1746" alt="Screenshot 2021-11-06 at 12 31 10" src="https://user-images.githubusercontent.com/26833433/140608327-9c0c0baf-a0ee-4551-a4ff-5ee26b155c51.png">

Good luck and let us know if you have any other questions!

**Reply by rgaufman** [reply]
This is super helpful but also if I were to buy a GPU for machine learning, e.g. yolov8 or yolo-nas, is there any reason to get one with 24gb and more? Is there a table somewhere of combinations of image sizes and batch sizes and GPU ram requirements?

**Reply by ShAmoNiA** [reply]
Greeting @DSA101 
I think below table can help you.

<table><thead><tr><th>Batch Size</th><th>Image Size</th><th>Dataset Size (Size and Range)</th><th>VRAM Requirement (Approx.)</th></tr></thead><tbody><tr><td>2</td><td>416x416</td><td>Small (1-10k images)</td><td>3.0 GB</td></tr><tr><td>2</td><td>800x800</td><td>Small (1-10k images)</td><td>6.0 GB</td></tr><tr><td>8</td><td>512x512</td><td>Small (1-10k images)</td><td>5.5 GB</td></tr><tr><td>8</td><td>1024x1024</td><td>Small (1-10k images)</td><td>12.0 GB</td></tr><tr><td>1</td><td>512x512</td><td>Medium (10k-100k images)</td><td>4.0 GB</td></tr><tr><td>1</td><td>1024x1024</td><td>Medium (10k-100k images)</td><td>8.0 GB</td></tr><tr><td>4</td><td>512x512</td><td>Medium (10k-100k images)</td><td>9.0 GB</td></tr><tr><td>4</td><td>1024x1024</td><td>Medium (10k-100k images)</td><td>18.0 GB</td></tr><tr><td>1</td><td>512x512</td><td>Large (100k-1M images)</td><td>5.0 GB</td></tr><tr><td>1</td><td>1024x1024</td><td>Large (100k-1M images)</td><td>10.0 GB</td></tr><tr><td>4</td><td>512x512</td><td>Large (100k-1M images)</td><td>11.5 GB</td></tr><tr><td>4</td><td>1024x1024</td><td>Large (100k-1M images)</td><td>24.0 GB</td></tr></tbody></table>

**Reply by rgaufman** [reply]
Amazing, thank you!

**Reply by rgaufman** [reply]
Just a quick question, most examples I can find show batch sizes of something like 64. Many warn about using small batch sizes. The default in YoloV8 is also 16, which sounds like it would need 96GB of ram with a large data set and image size to use an optimal batch size?

**Reply by ShAmoNiA** [reply]
my bad! I didn't quite catch what you were asking.
you mean why YoloV8 used batch size = 16?

**Reply by rgaufman** [reply]
Yes but more importantly what effect does batch size number have on the resultant model’s accuracy. Is a model trained with batch size 16 or 64 more accurate than ones trained with batch size 1 or 2? By how much?

**Reply by ShAmoNiA** [reply]
Please check below link:
https://stats.stackexchange.com/questions/153531/what-is-batch-size-in-neural-network

Now, let's check the what is advantages and drawbacks when we use small batch size compared to large batch size.

1. Training Efficiency: Larger batch sizes can take advantage of parallel processing and hardware optimizations, leading to faster training times. If you have access to powerful hardware (e.g., GPUs) and want to maximize computational efficiency, using larger batch sizes can be beneficial.

2. Generalization: Smaller batch sizes, such as 1 or a few samples, can provide more frequent updates to the model's parameters and expose the model to a wider variety of training examples. This may help the model generalize better to new, unseen data. However, it can also introduce more noise into the training process.

3. Memory Constraints: Larger batch sizes require more memory to store the intermediate activations and gradients during training. If you have limited memory resources, using smaller batch sizes might be necessary to fit the data in memory.

4. Convergence Behavior: The choice of batch size can influence the convergence behavior of the model. In some cases, larger batch sizes might converge to a slightly better solution due to more stable gradient estimates, while smaller batch sizes might exhibit more erratic convergence behavior.

In conclusion, just know that, **Using a larger batch size does not necessarily guarantee better accuracy.**

And for "why YoloV8 used batch size = 16" :
Generally like above attitudes, by using larger value for batch size we will get these improvement:
1. Improved parallelism
2. Increased generalization
3. Stabilized gradients
